src/Proper/Plutus.hs

```diff
@@ -0,0 +1,508 @@
+{-# LANGUAGE TypeFamilies #-}
+
+--------------------------------------------------------------------------------
+-- Proper.Plutus encourages you to define a model before writing any Plutus
+--------------------------------------------------------------------------------
+module Proper.Plutus (
+  Proper (..),
+  IsProperty,
+  justDatumHash,
+  Formula (..),
+  CompiledObject (..),
+) where
+
+import Data.Functor.Identity (Identity)
+import Data.Kind (Type)
+import Data.List (notElem)
+import Data.Map.Lazy qualified as M
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String (fromString)
+import Data.Text (Text)
+import Hedgehog (
+  GenBase,
+  Group (..),
+  MonadGen,
+  MonadTest,
+  failure,
+  footnote,
+  footnoteShow,
+  forAll,
+  property,
+  success,
+  (===),
+ )
+import Hedgehog qualified
+import Hedgehog.Gen qualified as Gen
+import Ledger (
+  Datum (..),
+  DatumHash,
+  MintingPolicy,
+  PubKeyHash,
+  Redeemer (..),
+  ScriptError (..),
+  StakeValidator,
+  TxId (..),
+  TxInInfo (..),
+  TxOut (..),
+  TxOutRef (..),
+  Validator,
+  Value,
+  always,
+  applyMintingPolicyScript,
+  applyStakeValidatorScript,
+  applyValidator,
+  datumHash,
+  evaluateScript,
+ )
+import Ledger.Address (scriptHashAddress)
+import Plutus.V1.Ledger.Api (
+  DCert,
+  POSIXTimeRange,
+  StakingCredential,
+ )
+import Plutus.V1.Ledger.Contexts (
+  ScriptContext (..),
+  ScriptPurpose (Spending),
+  TxInfo (..),
+ )
+import Plutus.V1.Ledger.Scripts (
+  Context (..),
+  Script,
+  ValidatorHash,
+ )
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget (..))
+import PlutusCore.Evaluation.Machine.ExMemory (ExCPU (..), ExMemory (..))
+import PlutusTx (toBuiltinData)
+import PlutusTx.Builtins (
+  BuiltinData,
+ )
+import SAT.MiniSat (
+  Formula (..),
+  satisfiable,
+  solve_all,
+ )
+import Safe (
+  headMay,
+ )
+import Text.PrettyPrint (
+  Doc,
+  Style (lineLength),
+  colon,
+  hang,
+  int,
+  renderStyle,
+  style,
+  text,
+  vcat,
+  ($+$),
+  (<+>),
+ )
+import Text.Show.Pretty (ppDoc)
+import Prelude (
+  Bool (..),
+  Bounded (..),
+  Either (..),
+  Enum,
+  Eq,
+  Int,
+  Integer,
+  Maybe (..),
+  Ord,
+  Show (..),
+  String,
+  filter,
+  fmap,
+  mempty,
+  pure,
+  snd,
+  zip,
+  ($),
+  (&&),
+  (.),
+  (<$>),
+  (<=),
+  (<>),
+  (==),
+  (>>),
+ )
+
+--------------------------------------------------------------------------------
+-- Propositional logic is used to define two aspects of a model.
+-- The expected outcome of a test and the sets of properties which are valid in
+-- conjunction.
+--------------------------------------------------------------------------------
+
+class (Enum c, Eq c, Ord c, Bounded c, Show c) => IsProperty c
```

Comment:
Is this just a collection of constraints with no behaviours? If so, this should be a `type` alias, if anything.

Reply:
Yep! That's all this is doing. I think I've seen this pattern recommended by a sensible Haskeller before but I'm not sure where. How would the type alias approach improve the API? With how it  currently is you would implement your properties like so

```Haskell
instance Proper MyModel where
  ...
  data Property MyModel
    = SomeProperty
    | SomeOtherProperty
    ...
    deriving stock (Enum, Eq, Ord, Bounded, Show)
    ...
```

Reply:
It saves you writing them all out in a function like this though.

```Haskell
  genProperties ::
    MonadGen m =>
    GenBase m ~ Identity =>
    IsProperty (Property model) =>
    model ->
    m (Set (Property model))
  genProperties _ = genGivenFormula logic
```

How could we do this better?

Reply:
```haskell
type Proper (a :: Type) = (Enum a, Eq a, Ord a, Bounded a, Show a)
```

Followed by

```haskell
genProperties :: 
   forall (model :: Type) . 
   Proper model => 
   model -> 
   Gen (Set (Property model))
```

We don't even need that rigidity constraint there, since `Gen` will do. FWIW, I'm _more_ in favour of just spelling out every constraint instead of using a type alias (which I am generally against), but lawless type classes are even worse, and that one doesn't even have methods.